modules/mqtt/mqtt-options.c

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright (c) 2021 One Identity
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "mqtt-options.h"
+#include "syslog-ng.h"
+
+#include <MQTTClient.h>
+#include <string.h>
+#include <stddef.h>
+
+
+#define DEFAULT_ADDRESS "tcp://localhost:1883"
+#define DEFAULT_KEEPALIVE 60
+#define DEFAULT_QOS 0
+
+
+
+static gboolean
+_validate_protocol(const gchar *address)
+{
+  const gchar *valid_type[] = {"tcp", "ssl", "ws", "wss"};
+  gint i;
+
+  for (i = 0; i < G_N_ELEMENTS(valid_type); ++i)
+    if (strncmp(valid_type[i], address, strlen(valid_type[i])) == 0)
+      return TRUE;
+
+  return FALSE;
+}
+
+static gboolean
+_validate_address(const gchar *address)
+{
+  if (strstr(address, "://") == NULL)
+    return FALSE;
+
+  if (!_validate_protocol(address))
+    return FALSE;
+
+  return TRUE;
+}
+
+static void
+_mqtt_internal_log(enum MQTTCLIENT_TRACE_LEVELS level, gchar *message)
+{
+  if (level >= MQTTCLIENT_TRACE_ERROR)
+    {
+      msg_error("MQTT error", evt_tag_str("error_message", message));
+      return;
+    }
+
+  msg_trace("MQTT debug", evt_tag_str("message", message));
+}
+
+void
+mqtt_client_options_defaults(MQTTClientOptions *self)
+{
+  self->address = g_strdup(DEFAULT_ADDRESS);
+  self->keepalive = DEFAULT_KEEPALIVE;
+  self->qos = DEFAULT_QOS;
+
+  self->ssl_version = MQTT_SSL_VERSION_DEFAULT;
+  self->peer_verify = TRUE;
+  self->use_system_cert_store = FALSE;
+
+  MQTTClient_setTraceCallback(_mqtt_internal_log);
+  MQTTClient_setTraceLevel(MQTTCLIENT_TRACE_PROTOCOL);
```

Comment:
These 2 calls are global within a single syslog-ng instance, so we can't configure them separately for each MQTT source/destination instance.

For this reason, I think they don't belong to `MQTTClientOptions`. Maybe they could go into `mqtt_module_init()`?!